crates/fluvio-spu/src/services/public/produce_handler.rs

```diff
@@ -212,48 +212,48 @@ async fn handle_produce_partition(
     }
 }
 
-async fn smartmodule_ctx(
-    sm_invocations: Vec<SmartModuleInvocation>,
+async fn apply_smartmodules(
+    partition_request: &mut PartitionProduceData<RecordSet<RawRecords>>,
+    smartmodules: &[SmartModuleInvocation],
     api_version: i16,
+    leader_state: &SharedFileLeaderState,
     ctx: &DefaultSharedGlobalContext,
-) -> Result<Option<SmartModuleContext>> {
-    SmartModuleContext::try_from(sm_invocations, api_version, ctx)
-        .await
-        .map_err(|error_code| {
-            warn!("smartmodule context init failed: {:?}", error_code);
-            anyhow!("smartmodule context init failed: {}", error_code)
-        })
-}
+) -> Result<(), ErrorCode> {
+    let Some(mut sm_ctx) =
+        SmartModuleContext::try_from(smartmodules.to_vec(), api_version, ctx).await?
+    else {
+        return Ok(());
+    };
+
+    sm_ctx.look_back(leader_state).await?;
 
-// #[cfg(feature = "smartengine")]
-fn apply_smartmodules_for_partition_request(
-    partition_request: &mut PartitionProduceData<RecordSet<RawRecords>>,
-    sm_chain_instance: &mut SmartModuleChainInstance,
-    ctx: &DefaultSharedGlobalContext,
-) -> Result<()> {
     let records = &partition_request.records;
     let batches = &records.batches;
 
     let mut batches = ProduceBatchIterator::new(batches);
 
     let sm_result = match process_batch(
-        sm_chain_instance,
+        sm_ctx.chain_mut(),
         &mut batches,
         std::usize::MAX,
         ctx.metrics().chain_metrics(),
     ) {
         Ok((result, sm_runtime_error)) => {
             if let Some(error) = sm_runtime_error {
-                return Err(anyhow!("smartmodule runtime error: {error}"));
```

Comment:
Instead of trying to revert changes from moving to Anyhow, let's try to find out why this was not caught since we may similar problem elsewhere.